Source/Core/VideoBackends/D3D12/D3DState.cpp

```diff
@@ -272,42 +272,40 @@ D3D12_BLEND GetBlendingAlpha(D3D12_BLEND blend)
   }
 }
 
-D3D12_BLEND_DESC StateCache::GetDesc12(BlendState state)
+D3D12_BLEND_DESC StateCache::GetDesc12(BlendingState state)
 {
-  if (!state.blend_enable)
-  {
-    state.src_blend = D3D12_BLEND_ONE;
-    state.dst_blend = D3D12_BLEND_ZERO;
-    state.blend_op = D3D12_BLEND_OP_ADD;
-    state.use_dst_alpha = false;
-  }
+  D3D12_BLEND_DESC desc = {};
+  desc.AlphaToCoverageEnable = FALSE;
+  desc.IndependentBlendEnable = FALSE;
 
-  D3D12_BLEND_DESC blenddc = {FALSE,  // BOOL AlphaToCoverageEnable;
-                              FALSE,  // BOOL IndependentBlendEnable;
-                              {
-                                  state.blend_enable,   // BOOL BlendEnable;
-                                  FALSE,                // BOOL LogicOpEnable;
-                                  state.src_blend,      // D3D12_BLEND SrcBlend;
-                                  state.dst_blend,      // D3D12_BLEND DestBlend;
-                                  state.blend_op,       // D3D12_BLEND_OP BlendOp;
-                                  state.src_blend,      // D3D12_BLEND SrcBlendAlpha;
-                                  state.dst_blend,      // D3D12_BLEND DestBlendAlpha;
-                                  state.blend_op,       // D3D12_BLEND_OP BlendOpAlpha;
-                                  D3D12_LOGIC_OP_NOOP,  // D3D12_LOGIC_OP LogicOp
-                                  state.write_mask      // UINT8 RenderTargetWriteMask;
-                              }};
-
-  blenddc.RenderTarget[0].SrcBlendAlpha = GetBlendingAlpha(blenddc.RenderTarget[0].SrcBlend);
-  blenddc.RenderTarget[0].DestBlendAlpha = GetBlendingAlpha(blenddc.RenderTarget[0].DestBlend);
-
-  if (state.use_dst_alpha)
-  {
-    blenddc.RenderTarget[0].SrcBlendAlpha = D3D12_BLEND_ONE;
-    blenddc.RenderTarget[0].DestBlendAlpha = D3D12_BLEND_ZERO;
-    blenddc.RenderTarget[0].BlendOpAlpha = D3D12_BLEND_OP_ADD;
-  }
+  D3D12_RENDER_TARGET_BLEND_DESC& tdesc = desc.RenderTarget[0];
+  tdesc.BlendEnable = state.blendenable;
 
-  return blenddc;
+  if (state.colorupdate)
+    tdesc.RenderTargetWriteMask = D3D12_COLOR_WRITE_ENABLE_RED | D3D12_COLOR_WRITE_ENABLE_GREEN |
+                                  D3D12_COLOR_WRITE_ENABLE_BLUE;
+  else
+    tdesc.RenderTargetWriteMask = 0;
+  if (state.alphaupdate)
+    tdesc.RenderTargetWriteMask |= D3D12_COLOR_WRITE_ENABLE_ALPHA;
+
+  static constexpr std::array<D3D12_BLEND, 8> src_factors = {
+      {D3D12_BLEND_ZERO, D3D12_BLEND_ONE, D3D12_BLEND_DEST_COLOR, D3D12_BLEND_INV_DEST_COLOR,
+       D3D12_BLEND_SRC1_ALPHA, D3D12_BLEND_INV_SRC1_ALPHA, D3D12_BLEND_DEST_ALPHA,
+       D3D12_BLEND_INV_DEST_ALPHA}};
+  static constexpr std::array<D3D12_BLEND, 8> dst_factors = {
+      {D3D12_BLEND_ZERO, D3D12_BLEND_ONE, D3D12_BLEND_SRC_COLOR, D3D12_BLEND_INV_SRC_COLOR,
+       D3D12_BLEND_SRC1_ALPHA, D3D12_BLEND_INV_SRC1_ALPHA, D3D12_BLEND_DEST_ALPHA,
+       D3D12_BLEND_INV_DEST_ALPHA}};
+
+  tdesc.SrcBlend = src_factors[state.srcfactor];
```

Comment:
Really minor but maybe move the source blends under the static array definition?